### DISCUSSION TITLE
SSH Keys

### DISCUSSION THREAD

**Discussion by larena1** [dani-garcia/vaultwarden#5158] [General]
https://github.com/bitwarden/server/pull/4575

As this will be merged soon, is it planned to add support for SSH Keys to Vaultwarden as well?

@dani-garcia @BlackDex

**Comment by BlackDex** [comment]
Looks simple enough as far as i can tell. But i have not validated it yet.

**Reply by larena1** [reply]
It has been merged into main now so maybe you could take a look?

**Reply by BlackDex** [reply]
The main other issue is the web-vault. There is no support yet for newer versions then 2024.6.2 in Vaultwarden yet. And it requires a specific version number to be used on the server side for it to work.

**Reply by larena1** [reply]
Do newer web versions require some new endpoints or why is that exactly?

**Reply by BlackDex** [reply]
The manager role has disappeared. 
I'm working on a quick fix to use the custom role and have it backwards compatible in the database instead of migrating everything back and forward which will b cause issues somewhere in the end.

**Reply by larena1** [reply]
It seems you are more or less maintaining Vaultwarden alone at this point and @dani-garcia has stopped contributing. Would be a shame if Vaultwarden was abandoned but it feels a bit like it already.

**Reply by liszca** [reply]
> Looks simple enough as far as i can tell. But i have not validated it yet.

Tried to save a ssh private key to a Custom Field of type "Hidden". Result was its not saved with the Newlines, if that's fixed or a type "Hidden Multi line" is added that would be good enough.

Alternative why not having Notes hidden by default? Also "Card" seems to be suitable for that task, no Problems with Multi liners here.

**Comment by larena1** [comment]
The feature has now also been added to the clients. Do you know if there's more to do to support the most recent client versions than just the manager role? Otherwise at some point it could also happen that Vaultwarden cannot be used at all anymore with up to date clients when it's already behind 6 months now, right?

**Reply by BlackDex** [reply]
Since we use Vaultwarden our self, we would of course thrive to get these features in.
There are probably some more features added or changed on how the web-vault works, and maybe some clients.

Those are not yet tested by me at least.

**Comment by dani-garcia** [comment]
This has been implemented today with https://github.com/dani-garcia/vaultwarden/pull/5187. It's usable only on **desktop** clients built from source now as long as you update the version and enable the required feature flags on the server. 

The official release of the clients will be next month, if everything goes according to plan.

**Reply by svoop** [reply]
Perfect moment to say BIG THANKS for all your work on Vaultwarden @dani-garcia @BlackDex !!

**Comment by dytavan** [comment]
the desktop app is up to date but can't. see any option to add an ssh key I'm on mac nor on vaultwarden server or bitclient desktop

**Reply by f46b83ee9** [reply]
Same for me, I added ssh-key-vault-item,ssh-agent to the EXPERIMENTAL_CLIENT_FEATURE_FLAGS env variable.
Even with the latest macOS version it still doesn't seem to work and I don't see the new ssh key type.

**Reply by BlackDex** [reply]
Have you added the options?
https://github.com/dani-garcia/vaultwarden/blob/fd512300448995735961cbeedcbe4b684be1c5fd/.env.template#L353-L355

**Reply by dytavan** [reply]
> Have you added the options?
> 
> https://github.com/dani-garcia/vaultwarden/blob/fd512300448995735961cbeedcbe4b684be1c5fd/.env.template#L353-L355

yes I did it
![CleanShot 2024-12-13 at 09 19 45@2x](https://github.com/user-attachments/assets/237a49aa-85bb-4af7-b998-1754884a7c0a)

**Reply by BlackDex** [reply]
It works for me on Linux.
If I'm correct this feature doesn't work on all platforms, but not sure.

I at least see it works fine on Linux with the AppImage.

**Reply by quexten** [reply]
There is a bug in 2024.12.0 introduced by a last-minute security fix, that breaks the agent on Linux/Mac: https://github.com/bitwarden/clients/pull/12356
For this reason it looks like ssh-keys will be delayed until the january release.

That said, the list of currently unsupported platforms/protocols:
- Snap
- Flatpak
- Windows Pageant (Putty/Filezilla)
- Windows Subsystem for Linux

All other platforms are expected to work.

**Reply by Faabvk** [reply]
Ah man I thought I was going crazy trying to figure out why it wasn't working for me.
From Admin > Diagnostics > Support string:
```
  "experimental_client_feature_flags": "fido2-vault-credentials,ssh-key-vault-item,ssh-agent",
```

And from my client
```
Version 2024.12.0
SDK 'main (edee180)'
Shell 32.1.1
Renderer 128.0.6613.137
Node 20.17.0
Architecture x64
```

**Reply by jradwan** [reply]
What is this supposed to look like in the UI after enabling the options? I've got the options added and the latest desktop client (on Windows 11) but can't find anywhere would I would add or manage SSH keys. I can't find any screenshots or documentation that describe how this is supposed to work. Can someone nudge me in the right direction?

**Reply by rushout912** [reply]
> What is this supposed to look like in the UI after enabling the options? I've got the options added and the latest desktop client (on Windows 11) but can't find anywhere would I would add or manage SSH keys. I can't find any screenshots or documentation that describe how this is supposed to work. Can someone nudge me in the right direction?

Go to settings and choose the "Enable SSH Agent" option :

![SSH Agent](https://github.com/user-attachments/assets/d27e374d-636e-407d-abcf-7c6d82366b32)
If you also enabled the ` ssh-key-vault-item`  flag, you should  see the new SSH Key item type :

![SSH Key Item Type](https://github.com/user-attachments/assets/38e5a91d-6985-4474-9fb5-1599cf8666d0)

**Reply by jradwan** [reply]
Thanks. I must be doing something wrong because I don't see that under Settings or the new Type. Running 1.32.7 in a Docker container on Synology. I'll keep poking ...

**Reply by jradwan** [reply]
Ok, maybe there's some sort of timing issue for the clients to "reconnect" and notice the config change? Because now suddenly I see the option in Settings and under Types. Thank you!

**Reply by ghost** [reply]
it would be nice to also support android (Termux)
https://github.com/termux/termux-app/issues/4289#issuecomment-2524947365

**Comment by WRW001** [comment]
Hello! I have my desktop client updated to 2024.12.0, and I also have the EXPERIMENTAL_CLIENT_FEATURE_FLAGS configured. And I can use it generate new ssh-keys. But I am having issues with the key algorithm, the only algorithm supported is ed25519, it seems that there is no chance  generate a RSA key or an ECDSA key, and also there is no chance to import any RSA/ECDSA key. Any solutions?

**Reply by dytavan** [reply]
with which os ?

**Reply by WRW001** [reply]
Windows11 23h2 and 24h2

**Reply by WRW001** [reply]
The server is running Ubuntu 24.04

**Reply by BlackDex** [reply]
There are some known issue with keys if I'm correct. Not all types are supported.
Some password protected keys for example do not work and maybe some others.

Please keep in mind that this is a client feature, and Vaultwarden only stores this and reports back to the clients the server supports this feature. How this is all handled is fully client-side.

**Reply by dytavan** [reply]
so only mac won't work ?

**Reply by WRW001** [reply]
@BlackDex Thanks for the lightning-fast reply. One thing to be clear: these non-ed25519 keys are not encrypted. I'm really confused right now. I have searched the client code and found no clues about this issue. The client does not have a limited key algorithm at all.

**Reply by BlackDex** [reply]
@quexten Maybe you can clear things up here :)

**Reply by WRW001** [reply]
@davidyannick86 Have you tried rebooting your server? I have the SSH-key option available after I rebooted my server.

**Reply by f46b83ee9** [reply]
Using Windows client it works, but not with the macOS one even after rebooting Vaultwarden,

**Reply by dytavan** [reply]
I've restarted the server yes

**Reply by f46b83ee9** [reply]
After disconnecting / reconnecting to Vaultwarden it works. Thanks a lot :)

**Reply by quexten** [reply]
> and also there is no chance to import any RSA/ECDSA key. Any solutions?

RSA keys are supported for import. However, some client (for instance putty) generate non-standard encodings (f.e using a 64 linewidth instead of 70), which needs an upstream dependency to be supported. Importing keys generated from openssh works though - **even password protected ones**.

@WRW001 which client did you generate/export your keys in?

That said, is there a good use-case you have for RSA/EcDSA? Both are cryptographically sub-par, compared to ed25519, and have both a higher risk of implementation vulnerability. For instance, the rsa crate, used in the ssh-agent, has unresolved timing sidechannels:  https://github.com/RustCrypto/RSA/pull/394 which could allow forging ssh signatures when you connect to a malicious server. That is to say, where possible, ed25519 should be used. I understand there is some legacy systems and embedded systems that do not yet support ed25519, so RSA is the alternative, discouraged, but still supported key type. I'm looking to hear about good use-cases though, because these strengthen the case for adding RSA to the generator instead of just allowing import.

**Reply by WRW001** [reply]
@quexten Thanks for your reply. These non-ED25519 keys were generated by SecureCRT and were converted to OpenSSH format. 
> (f.e using a 64 linewidth instead of 70)

I have those keys checked and they are 64 linewidth!  I also have had a try for a 70 linewidth RSA key,  It was successful!
 
Yes. The ED25519 is cryptographically superior to RSA and ECDSA, but I do have some systems that do not yet support ed25519. So I have to use RSA and ECDSA.

**Reply by WRW001** [reply]
@quexten 
> That said, is there a good use-case you have for RSA/EcDSA?

RFC 6187
X.509 Certificates can be used for both client and server authentication. This is the scenario in which RSA and ECDSA have widely been used.

**Reply by rushout912** [reply]
> > and also there is no chance to import any RSA/ECDSA key. Any solutions?
> 
> RSA keys are supported for import. However, some client (for instance putty) generate non-standard encodings (f.e using a 64 linewidth instead of 70), which needs an upstream dependency to be supported. Importing keys generated from openssh works though - **even password protected ones**.
> 
> @WRW001 which client did you generate/export your keys in?
> 
> That said, is there a good use-case you have for RSA/EcDSA? Both are cryptographically sub-par, compared to ed25519, and have both a higher risk of implementation vulnerability. For instance, the rsa crate, used in the ssh-agent, has unresolved timing sidechannels: [RustCrypto/RSA#394](https://github.com/RustCrypto/RSA/pull/394) which could allow forging ssh signatures when you connect to a malicious server. That is to say, where possible, ed25519 should be used. I understand there is some legacy systems and embedded systems that do not yet support ed25519, so RSA is the alternative, discouraged, but still supported key type. I'm looking to hear about good use-cases though, because these strengthen the case for adding RSA to the generator instead of just allowing import.

Sorry for bumping into discussion, is password encrypted ssh key import supported in current client versions (2024.12.0)? When I tried to import such keys, client asked me to enter the password for the SSH key, I typed the password, and then client said the SSH key is invalid. I checked the key's linewidth and I'm sure it has 70 linewidth, which is OpenSSH format.

**Reply by WRW001** [reply]
@rushout912 Are you using the ECDSA key? It is currently totally unsupported. A 70-linewidth RSA or ED25519 key seems OK.

**Reply by quexten** [reply]
> Sorry for bumping into discussion, is password encrypted ssh key import supported in current client versions (2024.12.0)? When I tried to import such keys, client asked me to enter the password for the SSH key, I typed the password, and then client said the SSH key is invalid. I checked the key's linewidth and I'm sure it has 70 linewidth, which is OpenSSH format.

I'd also recommend creating a GitHub issue on the upstream bitwarden/clients. It would especially be helpful to have test keys, or the application they are generated in.

**Reply by WRW001** [reply]
@quexten A new issue has been filed on the upstream bitwarden/clients.
https://github.com/bitwarden/clients/issues/12412

**Comment by foux** [comment]
I've enabled the SSH-Agent on macOS, but it won't look to bitwarden keys. Is there a next step to enable the ssh-agent?
Thanks

**Reply by fdisamuel** [reply]
I did the same on Windows Desktop Version 2024.12.1, neither do I know how to instruct windows terminal or windows ssh-agent to look for keys in bitwarden vault.. Any help appreciated!

**Reply by quexten** [reply]
The official instructions are not out yet because this is not released yet on upstream Bitwarden. But for windows you can use the same instructions that 1Password provides for their agent, then restart Bitwarden desktop.

For mac and linux you need to put `SSH_AUTH_SOCK=/{MY_USER_HOME_DIR}/.bitwarden-ssh-agent.sock` (putting in your home directory, in your environment variables.

**Reply by fdisamuel** [reply]
Thank you soo much @quexten, by following the guide on 1Password Developer Center, I was able to set up my Bitwarden Desktop ssh-agent integration! It works like a charm now 😊

**Reply by foux** [reply]
> For mac and linux you need to put `SSH_AUTH_SOCK=/{MY_USER_HOME_DIR}/.bitwarden-ssh-agent.sock` (putting in your home directory, in your environment variables.

Thanks @quexten. I've got the ssh-agent activated in bitwarden preferences, but no `~/.bitwarden-ssh-agent.sock` on macOS. Any idea why?
Thanks

**Reply by quexten** [reply]
Yeah, there was a security fix that caused a bug which causes the agent to crash if the aforementioned socket is not already present. That's the reason the release got delayed to january.

Either you can download a dev build from bitwarden/clients's CI pipeline, or you can run:

`touch ~/.bitwarden-ssh-agent.sock`, and restart bitwarden desktop.

**Reply by foux** [reply]
Touching the sock file didn't work for me, I'll try with the dev builds, thanks.

**Reply by larena1** [reply]
@quexten are there any news regarding PuTTY agent support or is this not being worked on?

**Reply by quexten** [reply]
It is not being actively worked on at the moment.
Pageant seems like a [horrific path](https://github.com/ndbeals/winssh-pageant/blob/master/pageant/run_windows.go) to go down.

Luckily there is tools bridging this gap, so if you need putty, you can use something like: https://github.com/ndbeals/winssh-pageant

**Reply by ghost** [reply]
does bitwarden ssh  agent coexist with micorsoft openssh's built-in agent, which supports yubikeys, or do I have to disable that to use bitwarden ssh agent?

**Reply by Delta1977** [reply]
It coexists. I use both.

**Reply by ghost** [reply]
@Delta1977  can you run them at the same time?
I use https://github.com/buptczq/WinCryptSSHAgent but it says open \\.\pipe\openssh-ssh-agent: Access is denied when bitwarden is open.

**Reply by Delta1977** [reply]
I don't know. I installed sublimes ssh described at yubikey'a support page and for Bitwarden I installed only the 2024.12 package for windows. That was all. If there is a fallback I didn't know as for some targets i have keys in Bitwarden , for some in yubikey and for some in .ssh but no key for the same target in multiple storage.

**Reply by foux** [reply]
@quexten I've updated to version `2025.1.0`, but the SSH auth sock still isn't working on macOS (even when creating `~/.bitwarden-ssh-agent.sock` beforehand). Is that expected? Thanks

**Reply by quexten** [reply]
Is this on the mac store version or the .dmg from the website?

**Reply by foux** [reply]
The store version

**Reply by quexten** [reply]
The store version currently does not support the ssh agent (Still need to write a PR to remove the setting from the menu). (The store version does sandboxing which needs special handling which I've not had the time to look into yet).

**Reply by foux** [reply]
So we have to choose between biometric unlock and SSH agent? (If I remember correctly the DMG version doesn't have biometric unlock)

**Reply by foux** [reply]
I've made a mistake : it's the Safari extension we loos if we use the DMG app. That's kind of difficult

**Reply by foux** [reply]
I've noticed another issue : if we close the main window, the agent won't work anymore : we won't have the popup to accept the use of the SSH key. Even after reopening the window, it won't ask to accept the key anymore. The only solution is to quit Bitwarden and reopen it

**Reply by quexten** [reply]
Thank you for the detailed report, I've raised this issue internally now. The fix probably won't make it to the next version but certainly is high priority. For now, does minimizing work?

**Reply by foux** [reply]
Well sorry but it seems I've added too many conditions to the issue. Just trying two times to use a key doesn't work, it only works on the first time (for the same key or another) : 
![CleanShot 2025-01-15 at 16 13 23](https://github.com/user-attachments/assets/35bbb1b7-9d4c-4172-83fd-a9979628e793)

Closing the window actually works, it will reopen it to ask access (but only the first time) : 
![CleanShot 2025-01-15 at 16 16 06](https://github.com/user-attachments/assets/6d67ce1d-b610-410b-833a-9290728de84c)


Sorry for the wrong first conclusions.

[edit : ] edited to hide instance URL and email from the GIFs

**Reply by mil1i** [reply]
Hm yeah. The mac-store version I think is a big deal. DMG installer doesn't support biometric unlock, on top of the loss of Safari support.

After migrating to the DMG installer, it did work after adding:

```shell
export SSH_AUTH_SOCK="${HOME}/.bitwarden-ssh-agent.sock"
```

Although, Bitwarden seemed to already be creating an `agent.sock` elsewhere. I suppose from above `${HOME}/.bitwarden-ssh-agent.sock` is a "vaultwarden" specific implementation vs the other being the MacStore's sandboxed location. The one created in `~/Library/Containers/com.bitwarden.desktop/Data` was being created by the MacStore version as well, but setting the `SSH_AUTH_SOCK` to point to it still didn't work.

```shell
❯ fd agent.sock -H
.bitwarden-ssh-agent.sock
Library/Containers/com.bitwarden.desktop/Data/.bitwarden-ssh-agent.sock
```

Was able to validate with `ssh-add -l` and then ssh'ing in. I didn't run into any issues with first-time run with the Bitwarden Desktop GUI closed out.

What about ssh git-sign? Is that also available, or on the roadmap for this integration like 1Password does?

```toml
[gpg "ssh"]
        program = /Applications/1Password.app/Contents/MacOS/op-ssh-sign
        allowedSignersFile = "~/.ssh/allowed_signers"
```

**Reply by quexten** [reply]
I don't know what 1password does with their `op-ssh-sign` binary. SSH signing of git commits works out of the box without any extra binary. The docs are not out yet, but my git config looks like this:

```
[user]
	name = Bernd Schoolmann
	email = mail@quexten.com
        signingkey = "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIJ2N+qs9c7ImW97etkOGNflQkb1lzMilnUj+3NharB/Z"
[commit]
        gpgsign = true

[gpg]
	format = ssh
```

> I suppose from above ${HOME}/.bitwarden-ssh-agent.sock is a "vaultwarden" specific implementation 

Vaultwarden implements nothing for the agent, it implements vault item storage only. So it shouldn't affect anything.

**Reply by foux** [reply]
> DMG installer doesn't support biometric unlock

No, I was wrong when I said this. You still have biometric unlock with the DMG. You only loose Safari integration

**Reply by lionep** [reply]
I have exactly the same behavior with the second time request to ssh-agent : first time is showing the "Authorise popup", and second time is just making my ssh client hanging.  

I've launched bitwarden client (DMG installed) from command line, and got this line 

```
[SSH Agent] Confirming request from application: ssh
```

only on first request.

So maybe it's an issue with the bitwarden macos desktop client ? Closing bitwarden window does not work for me, I have to completely kill bitwarden client, and relaunch it.

**Reply by quexten** [reply]
Are you by chance using ssh agent forwarding? So far no luck reproducing this internally.

**Reply by foux** [reply]
> Are you by chance using ssh agent forwarding? So far no luck reproducing this internally.

Yes, in my case I am. I'll try without

**Reply by foux** [reply]
@quexten I confirm I haven't got the issue when removing agent forwarding

**Reply by quexten** [reply]
In that case it is fixed by this PR: https://github.com/bitwarden/clients/pull/12371

**Reply by foux** [reply]
Thanks a lot, and sorry for the time wasted testing a wrong issue.

**Reply by lionep** [reply]
Thanks! Same for me, agent forwarding is enabled

**Reply by svoop** [reply]
> For mac and linux you need to put `SSH_AUTH_SOCK=/{MY_USER_HOME_DIR}/.bitwarden-ssh-agent.sock` (putting in your home directory, in your environment variables.

Some terminal apps like iTerm or Wave SSH directly without sourcing `.zprofile` where this environment variable might be set. Therefore, a maybe better way to set the socket is in `~/.ssh/config`:

```
Host *
    IdentityAgent ~/.bitwarden-ssh-agent.sock
```

This way, you can also use Bitwarden for some hosts and another agent of other hosts. (Don't really see any reason why, but hey, you could. :smile: )

**Reply by smile2008** [reply]
Hi, on macOS 
I done
export SSH_AUTH_SOCK="${HOME}/.bitwarden-ssh-agent.sock"
and all seems to be ok:

ssh-add -l                                              
256 SHA256:76pY7ubsXPBs65VcaiGB5FHE5f1aFpQuw8He1fz0CtY Test server (ED25519)

When I try to ssh, I got popup in client, hit allow button and.....
got this error:
sign_and_send_pubkey: signing failed for ED25519 "Test server" from agent: agent refused operation

Please tell me how repair this(

**Comment by nyawox** [comment]
i want to thank you for all the work put on this feature, for both upstream and vaultwarden contributors!
I'm daily driving this (although admitelly i had to run a "patched" bitwarden desktop with cherry-picked prs, packaged desktop_proxy for nix and replaced nord with custom theme)
it's already been few weeks and i didn't have any issue on the backend until now, running flawlessly. 
the only nice to have qol feature i see lacking today is on the client side, which might be possible after https://redirect.github.com/bitwarden/clients/pull/12371
but other than that, i think it's safe to say it's mature enough to rely on as ssh key storage, or even better, you could add `pam_rssh` to avoid reusing memorable user passwords across deployed machines.

**Comment by arpanghosh8453** [comment]
Hello! This looks like an amazing feature. I do see the SSH key subtype in the browser extension, but can't add any there. Inside the add option, there is to SSH key field. 

I have added the following in the experimental flag. Server is latest docker container. 

` EXPERIMENTAL_CLIENT_FEATURE_FLAGS=fido2-vault-credentials,ssh-key-vault-item,ssh-agent`

I am using Bitwarden official clients Version: 2025.1.0 for Extension as well as Desktop app. 

**Web app (No SSH key addition possible):**

![image](https://github.com/user-attachments/assets/f5cf3b1b-70bd-4442-bc8c-37ba30133a00)

**Browser Extension shows the SSH keys in type:**

![image](https://github.com/user-attachments/assets/149a77dc-81b0-4457-9542-d4d4e55bdcf0)

**Browser Extension unable to add any SSH keys:**

![image](https://github.com/user-attachments/assets/c93b125f-de69-4984-b757-3c7afb3c580c)

What I am missing here? @dani-garcia @BlackDex

**Reply by BlackDex** [reply]
I think patients. Most of the clients which have full support have just been released. And some clients cache the response of the config endpoint for a while. 

The web-vault doesn't support it yet unless you use the trading i tagged releases.

**Reply by quexten** [reply]
Generating ssh keys in the extension is only supported from 2025.1.1 onwards, same thing goes for the web vault.

**Reply by arpanghosh8453** [reply]
Thank you @BlackDex . I used the Windows store earlier which installed 2025.1.0. Now I updated that to 2025.1.2 with manual install and it looks like working in the Desktop client. Thank you for the superfast reply! Love Vaultwarden!

**Comment by smile2008** [comment]
If I have several ssh keys in my vault, I often got (with macOS, Windows clients ant their ssh-agents:

`_Received disconnect from my_ip_or_host_name port 22:2: Too many authentication failures._`

How I can specify which key to use? Is there such way()?

**Comment by prljav** [comment]
when is this getting officially released?

**Comment by prljav** [comment]
NVM, started working by itself
i cannot get this to work, i set my env variable to:
EXPERIMENTAL_CLIENT_FEATURE_FLAGS=fido2-vault-credentials,ssh-key-vault-item,ssh-agent
and in the extension, there is no mention of ssh keys
in the desktop app, i dont even see the option to enable them (desktop version 2025.2.0)

**Reply by gleupold** [reply]
Hi @prljav ,
i ran into that problem too. 
Than i opened the Web Frontend and saw  the ability to add the ssh-key. Next i logout from Desktop client and signed in again, voila: The options showed up.

**Reply by kekorder** [reply]
Anyone know when this will be supported by default on the server?
